R/estimateDiversity.R

```diff
@@ -453,24 +453,34 @@ setMethod("estimateFaith", signature = c(x="TreeSummarizedExperiment", tree="mis
     }
     # Determine the quantile point.
     quantile_point <- quantile(max(mat), quantile)
-    
     # Tabulate the arithmetic abundance classes. Use the same classes
     # for all samples for consistency    
     cutpoints <- c(seq(0, quantile_point, length=num_of_classes), Inf)
-    
-    # Check skewness of the abundance classes for each sample
-    r <- apply(mat, 2, function(x) {
-        .calc_skewness(table(cut(x, cutpoints)))
-    })
-    
+    # Calculates sample-wise frequencies. How many taxa in each interval?
+    freq_table <- table(c(cut(mat, cutpoints)), col(mat))
+    # Calculates the skewness of frequency table. Returns skewness for each sample
+    r <- .calc_skewness(freq_table)
```

Comment:
I think here is a wrong result returned.

```r
> freq_table <- table(c(cut(mat, cutpoints)), col(mat))
> mia:::.calc_skewness(freq_table)
[1] 2.733111 2.140527 2.691113
```

```r
> mia:::.calc_skewness(apply(mat, 2, function(x) {
+     table(cut(x, cutpoints))
+ }))
[1] 6.490312 5.216445 6.490942
```

I think you need to include a test for correct values.

Reply:
I added tests. However, I got the same result from your example. I also tested `.calc_skewness` with `microbiome'`s value's and they were the same

Reply:
Hmm, I get the following result

```r
> table(c(cut(mat, cutpoints)), col(mat))
    
      1  2  3
  1  19 16 28
  2   3  6  5
  3   1  3  1
  4   1  2  2
  6   1  2  0
  8   0  1  1
  15  1  0  0
  17  0  3  0
  20  1  0  0
  21  1  0  0
  49  0  0  1
```

but the alternative is of course different

```r
> apply(mat, 2, function(x) {
+     table(cut(x, cutpoints))
+ })
             B  C  D
(0,2.53]    19 16 28
(2.53,5.06]  3  6  5
(5.06,7.59]  1  3  1
(7.59,10.1]  1  2  2
(10.1,12.7]  0  0  0
(12.7,15.2]  1  2  0
(15.2,17.7]  0  0  0
(17.7,20.2]  0  1  1
(20.2,22.8]  0  0  0
(22.8,25.3]  0  0  0
(25.3,27.8]  0  0  0
(27.8,30.4]  0  0  0
(30.4,32.9]  0  0  0
(32.9,35.4]  0  0  0
(35.4,38]    1  0  0
(38,40.5]    0  0  0
(40.5,43]    0  3  0
(43,45.6]    0  0  0
(45.6,48.1]  0  0  0
(48.1,50.6]  1  0  0
(50.6,53.1]  1  0  0
(53.1,55.7]  0  0  0
(55.7,58.2]  0  0  0
(58.2,60.7]  0  0  0
(60.7,63.3]  0  0  0
(63.3,65.8]  0  0  0
(65.8,68.3]  0  0  0
(68.3,70.9]  0  0  0
(70.9,73.4]  0  0  0
(73.4,75.9]  0  0  0
(75.9,78.4]  0  0  0
(78.4,81]    0  0  0
(81,83.5]    0  0  0
(83.5,86]    0  0  0
(86,88.6]    0  0  0
(88.6,91.1]  0  0  0
(91.1,93.6]  0  0  0
(93.6,96.2]  0  0  0
(96.2,98.7]  0  0  0
(98.7,101]   0  0  0
(101,104]    0  0  0
(104,106]    0  0  0
(106,109]    0  0  0
(109,111]    0  0  0
(111,114]    0  0  0
(114,116]    0  0  0
(116,119]    0  0  0
(119,121]    0  0  0
(121,124]    0  0  1
(124,Inf]    0  0  0
```

So this cannot be just happing in my session. `rowMeans` is definitly different with all the zeros in there. Can you check again?

The tests also don't pass for me locally, but they do on GitHub. This weird and probably dangerous

Reply:
```
> R.version
               _                                        
platform       x86_64-pc-linux-gnu                      
arch           x86_64                                   
os             linux-gnu                                
system         x86_64, linux-gnu                        
status         alpha                                    
major          4                                        
minor          1.0                                      
year           2021                                     
month          04                                       
day            26                                       
svn rev        80229                                    
language       R                                        
version.string R version 4.1.0 alpha (2021-04-26 r80229)
nickname                                                
> table(c(cut(mat, cutpoints)), col(mat))
             
               1  2  3
  (0,2.53]    19 16 28
  (2.53,5.06]  3  6  5
  (5.06,7.59]  1  3  1
  (7.59,10.1]  1  2  2
  (10.1,12.7]  0  0  0
  (12.7,15.2]  1  2  0
  (15.2,17.7]  0  0  0
  (17.7,20.2]  0  1  1
  (20.2,22.8]  0  0  0
  (22.8,25.3]  0  0  0
  (25.3,27.8]  0  0  0
  (27.8,30.4]  0  0  0
  (30.4,32.9]  0  0  0
  (32.9,35.4]  0  0  0
  (35.4,38]    1  0  0
  (38,40.5]    0  0  0
  (40.5,43]    0  3  0
  (43,45.6]    0  0  0
  (45.6,48.1]  0  0  0
  (48.1,50.6]  1  0  0
  (50.6,53.1]  1  0  0
  (53.1,55.7]  0  0  0
  (55.7,58.2]  0  0  0
  (58.2,60.7]  0  0  0
  (60.7,63.3]  0  0  0
  (63.3,65.8]  0  0  0
  (65.8,68.3]  0  0  0
  (68.3,70.9]  0  0  0
  (70.9,73.4]  0  0  0
  (73.4,75.9]  0  0  0
  (75.9,78.4]  0  0  0
  (78.4,81]    0  0  0
  (81,83.5]    0  0  0
  (83.5,86]    0  0  0
  (86,88.6]    0  0  0
  (88.6,91.1]  0  0  0
  (91.1,93.6]  0  0  0
  (93.6,96.2]  0  0  0
  (96.2,98.7]  0  0  0
  (98.7,101]   0  0  0
  (101,104]    0  0  0
  (104,106]    0  0  0
  (106,109]    0  0  0
  (109,111]    0  0  0
  (111,114]    0  0  0
  (114,116]    0  0  0
  (116,119]    0  0  0
  (119,121]    0  0  0
  (121,124]    0  0  1
  (124,Inf]    0  0  0
  > apply(mat, 2, function(x) {
+   table(cut(x, cutpoints))
+ })
             B  C  D
(0,2.53]    19 16 28
(2.53,5.06]  3  6  5
(5.06,7.59]  1  3  1
(7.59,10.1]  1  2  2
(10.1,12.7]  0  0  0
(12.7,15.2]  1  2  0
(15.2,17.7]  0  0  0
(17.7,20.2]  0  1  1
(20.2,22.8]  0  0  0
(22.8,25.3]  0  0  0
(25.3,27.8]  0  0  0
(27.8,30.4]  0  0  0
(30.4,32.9]  0  0  0
(32.9,35.4]  0  0  0
(35.4,38]    1  0  0
(38,40.5]    0  0  0
(40.5,43]    0  3  0
(43,45.6]    0  0  0
(45.6,48.1]  0  0  0
(48.1,50.6]  1  0  0
(50.6,53.1]  1  0  0
(53.1,55.7]  0  0  0
(55.7,58.2]  0  0  0
(58.2,60.7]  0  0  0
(60.7,63.3]  0  0  0
(63.3,65.8]  0  0  0
(65.8,68.3]  0  0  0
(68.3,70.9]  0  0  0
(70.9,73.4]  0  0  0
(73.4,75.9]  0  0  0
(75.9,78.4]  0  0  0
(78.4,81]    0  0  0
(81,83.5]    0  0  0
(83.5,86]    0  0  0
(86,88.6]    0  0  0
(88.6,91.1]  0  0  0
(91.1,93.6]  0  0  0
(93.6,96.2]  0  0  0
(96.2,98.7]  0  0  0
(98.7,101]   0  0  0
(101,104]    0  0  0
(104,106]    0  0  0
(106,109]    0  0  0
(109,111]    0  0  0
(111,114]    0  0  0
(114,116]    0  0  0
(116,119]    0  0  0
(119,121]    0  0  0
(121,124]    0  0  1
(124,Inf]    0  0  0
```

Reply:
```
> R.version
               _                           
platform       x86_64-w64-mingw32          
arch           x86_64                      
os             mingw32                     
system         x86_64, mingw32             
status                                     
major          4                           
minor          0.2                         
year           2020                        
month          06                          
day            22                          
svn rev        78730                       
language       R                           
version.string R version 4.0.2 (2020-06-22)
nickname       Taking Off Again            
> table(c(cut(mat, cutpoints)), col(mat))
    
      1  2  3
  1  19 16 28
  2   3  6  5
  3   1  3  1
  4   1  2  2
  6   1  2  0
  8   0  1  1
  15  1  0  0
  17  0  3  0
  20  1  0  0
  21  1  0  0
  49  0  0  1
> apply(mat, 2, function(x) {
+       table(cut(x, cutpoints))
+    })
             B  C  D
(0,2.53]    19 16 28
(2.53,5.06]  3  6  5
(5.06,7.59]  1  3  1
(7.59,10.1]  1  2  2
(10.1,12.7]  0  0  0
(12.7,15.2]  1  2  0
(15.2,17.7]  0  0  0
(17.7,20.2]  0  1  1
(20.2,22.8]  0  0  0
(22.8,25.3]  0  0  0
(25.3,27.8]  0  0  0
(27.8,30.4]  0  0  0
(30.4,32.9]  0  0  0
(32.9,35.4]  0  0  0
(35.4,38]    1  0  0
(38,40.5]    0  0  0
(40.5,43]    0  3  0
(43,45.6]    0  0  0
(45.6,48.1]  0  0  0
(48.1,50.6]  1  0  0
(50.6,53.1]  1  0  0
(53.1,55.7]  0  0  0
(55.7,58.2]  0  0  0
(58.2,60.7]  0  0  0
(60.7,63.3]  0  0  0
(63.3,65.8]  0  0  0
(65.8,68.3]  0  0  0
(68.3,70.9]  0  0  0
(70.9,73.4]  0  0  0
(73.4,75.9]  0  0  0
(75.9,78.4]  0  0  0
(78.4,81]    0  0  0
(81,83.5]    0  0  0
(83.5,86]    0  0  0
(86,88.6]    0  0  0
(88.6,91.1]  0  0  0
(91.1,93.6]  0  0  0
(93.6,96.2]  0  0  0
(96.2,98.7]  0  0  0
(98.7,101]   0  0  0
(101,104]    0  0  0
(104,106]    0  0  0
(106,109]    0  0  0
(109,111]    0  0  0
(111,114]    0  0  0
(114,116]    0  0  0
(116,119]    0  0  0
(119,121]    0  0  0
(121,124]    0  0  1
(124,Inf]    0  0  0
```

Reply:
This problem is related to 4.0

Reply:
Not exaclty. It is actually the behavior of `c` that changed. From the release notes

```
Using c() to combine a factor with other factors now gives a factor, an ordered factor when combining ordered factors with identical levels.
```

I fixed this in 70ddd9a5